File: TangHongyin/3sum_closet.py
class Solution(object):

    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        diff = 0xffffffff
        nums = sorted(nums)
        for i in range(len(nums)):
            low = i + 1
            high = len(nums) - 1
            while low < high:
                sum = nums[low] + nums[high] + nums[i]
                temp = sum - target
                if abs(temp) < abs(diff):
                    diff = temp
                if temp == 0:
                    return sum
                if temp < 0:
                    low += 1
                if temp > 0:
                    high -= 1
        return target + diff

if __name__ == '__main__':
    nums = [0,2,1,-3]
    solution = Solution()
    print solution.threeSumClosest(nums, 1)

File: Yandong/3SumCloset.cpp
class Solution {
public:
    /*固定一点，在其后面查找，变为2Sum Problem*/
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int res;
        bool first = true;
        for(int i=0;i<nums.size();i++){
            int start = i + 1;
            int end = nums.size() - 1;


            while(start < end){
                int sum = nums[i] + nums[start] + nums[end];

                if(first){
                    res = sum;
                    first = false;
                }
                else{
                    if(abs(res-target) > abs(sum-target)){
                        res = sum;
                    }
                }
                if(res == target)
                    return res;
                if(sum > target){
                    end --;
                }
                else{
                    start++;
                }
            }

        }
        return res;
    }
};
